Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -21,29 +21,28 @@ nonterminal symbol on the left-hand side of each rule.
 
 open Function
 
-universe uT uN in
 /-- Rule that rewrites a single nonterminal to any string (a list of symbols). -/
 @[ext]
-structure ContextFreeRule (T : Type uT) (N : Type uN) where
+structure ContextFreeRule (T : Type*) (N : Type) where
   /-- Input nonterminal a.k.a. left-hand side. -/
-  input : N
+  input : N -- `N` is not `Type*` because only finite types are useful,
+            -- and bundling (below) universe-polymophic types is ugly.
   /-- Output string a.k.a. right-hand side. -/
   output : List (Symbol T N)
 
 /-- Context-free grammar that generates words over the alphabet `T` (a type of terminals). -/
-structure ContextFreeGrammar.{uN,uT} (T : Type uT) where
+structure ContextFreeGrammar (T : Type*) where
   /-- Type of nonterminals. -/
```

Comment:
... it should rather be here
```suggestion
  /-- Type of nonterminals.

  Note that we restrict to `Type` because universe-polymorphism would be unnecessary: One may always
  restrict a context-free grammar to the finitely many nonterminal symbols that are referred to by
  its finitely many rules. -/
```